javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+            Function<? super T, Tuple3<? extends T1, ? extends T2, ? extends T3>> unzipper) {
+        Objects.requireNonNull(unzipper, "unzipper is null");
+        return iterator().unzip3(unzipper).map(
+                i1 -> TreeSet.ofAll(naturalComparator(), i1),
+                i2 -> TreeSet.ofAll(naturalComparator(), i2),
+                i3 -> TreeSet.ofAll(naturalComparator(), i3));
+    }
+
+    // TODO
+    @Override
+    default <U> TreeSet<Tuple2<T, U>> zip(Iterable<? extends U> that) {
+        Objects.requireNonNull(that, "that is null");
+        final Comparator<Tuple2<T, U>> tuple2Comparator = Tuple2.comparator(comparator(), naturalComparator());
+        return TreeSet.ofAll(tuple2Comparator, iterator().zip(that));
+    }
+
+    // TODO
+    @Override
+    default <U> TreeSet<Tuple2<T, U>> zipAll(Iterable<? extends U> that, T thisElem, U thatElem) {
+        Objects.requireNonNull(that, "that is null");
+        final Comparator<Tuple2<T, U>> tuple2Comparator = Tuple2.comparator(comparator(), naturalComparator());
+        return TreeSet.ofAll(tuple2Comparator, iterator().zipAll(that, thisElem, thatElem));
+    }
+
+    // TODO
+    @Override
+    default TreeSet<Tuple2<T, Long>> zipWithIndex() {
+        final Comparator<? super T> component1Comparator = comparator();
+        final Comparator<Tuple2<T, Long>> tuple2Comparator = (t1, t2) -> component1Comparator.compare(t1._1, t2._1);
+        return TreeSet.ofAll(tuple2Comparator, iterator().zipWithIndex());
+    }
+}
+
+interface BitSetModule {
+
+    int ADDRESS_BITS_PER_WORD = 6;
+    int BITS_PER_WORD = 64;
+
+    abstract class AbstractBitSet<T> implements BitSet<T> {
+
+        private static final long serialVersionUID = 1L;
+
+        final Function1<Integer, T> fromInt;
+        final Function1<T, Integer> toInt;
+
+        AbstractBitSet(Function1<Integer, T> fromInt, Function1<T, Integer> toInt) {
+            this.fromInt = fromInt;
+            this.toInt = toInt;
+        }
+
+        abstract int getWordsNum();
+
+        abstract long[] copyExpand(int wordsNum);
+
+        abstract long getWord(int index);
+
+        BitSet<T> createEmpty() {
+            return new BitSet1<>(fromInt, toInt, 0L);
+        }
+
+        BitSet<T> createFromAll(Iterable<? extends T> values) {
+            return createEmpty().addAll(values);
+        }
+
+        BitSet<T> fromBitMaskNoCopy(long[] elements) {
+            final int len = elements.length;
+            if (len == 0) {
+                return createEmpty();
+            }
+            if (len == 1) {
+                return new BitSet1<>(fromInt, toInt, elements[0]);
+            }
+            if (len == 2) {
+                return new BitSet2<>(fromInt, toInt, elements[0], elements[1]);
+            }
+            return new BitSetN<>(fromInt, toInt, elements);
+        }
+
+        private void setElement(long[] words, int element) {
+            final int index = element >> ADDRESS_BITS_PER_WORD;
```

Comment:
are we using the sign bit also for storage?
Do we need a `>>` or `>>>`?
Would this change if later we decide that negative indices mean indexing from the end somehow (not sure how it could be done in this case, though)?